File: 2219-84-873-length-of-longest-fibonacci-subsequence/2219-84-873-length-of-longest-fibonacci-subsequence.py
from typing import List

class Solution:
    def lenLongestFibSubseq(self, arr: List[int]) -> int:
        n = len(arr)
        dp = [[0] * n for _ in range(n)]
        mp = {x: i for i, x in enumerate(arr)}
        ans = 0

        for i in range(n):
            for j in range(i + 1, n):
                next_val = arr[i] + arr[j]
                if next_val in mp:
                    next_idx = mp[next_val]
                    dp[j][next_idx] = dp[i][j] + 1
                    ans = max(ans, dp[j][next_idx])

        return ans + 2 if ans > 0 else 0

File: 2219-84-873-length-of-longest-fibonacci-subsequence/2219-84-873-length-of-longest-fibonacci-subsequence.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lenLongestFibSubseq(int[] arr) {
        int n = arr.length;
        int[][] dp = new int[n][n];
        Map<Integer, Integer> mp = new HashMap<>();
        int ans = 0;

        for (int i = 0; i < n; i++) {
            mp.put(arr[i], i);
        }

        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int nextVal = arr[i] + arr[j];
                if (mp.containsKey(nextVal)) {
                    int nextIdx = mp.get(nextVal);
                    dp[j][nextIdx] = dp[i][j] + 1;
                    ans = Math.max(ans, dp[j][nextIdx]);
                }
            }
        }

        return ans > 0 ? ans + 2 : 0;
    }
}